core/src/main/scala/org/apache/gearpump/cluster/master/AppMasterLauncher.scala

```diff
@@ -0,0 +1,140 @@
+import InMemoryKVService._
+import MasterHAService._
+import org.apache.gearpump.cluster.MasterToAppMaster._
+import org.apache.gearpump.cluster.MasterToClient.{ReplayApplicationResult, ResolveAppIdResult, ShutdownApplicationResult, SubmitApplicationResult}
+import org.apache.gearpump.cluster.WorkerToAppMaster._
+import org.apache.gearpump.cluster._
+import org.apache.gearpump.cluster.appmaster.AppMasterDaemon
+import org.apache.gearpump.cluster.scheduler.{Resource, ResourceAllocation, ResourceRequest}
+import org.apache.gearpump.transport.HostPort
+import org.apache.gearpump.util.ActorSystemBooter._
+import org.apache.gearpump.util.Constants._
+import org.apache.gearpump.util._
+import org.slf4j.Logger
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success}
+import scala.concurrent.duration.Duration
+
+class AppMasterLauncher(appId : Int, executorId: Int, app : Application, jar: Option[AppJar], username : String, master : ActorRef, client: Option[ActorRef]) extends Actor {
+  private val LOG: Logger = LogUtil.getLogger(getClass, app = appId)
+
+  val scheduler = context.system.scheduler
+  val systemConfig = context.system.settings.config
+  val TIMEOUT = Duration(15, TimeUnit.SECONDS)
+  val appMaster = app.appMaster
+  val userConfig = app.conf
+
+  LOG.info(s"AppManager asking Master for resource for app $appId...")
+  master ! RequestResource(appId, ResourceRequest(Resource(1)))
+
+  def receive : Receive = waitForResourceAllocation
+
+  def waitForResourceAllocation : Receive = {
+    case ResourceAllocated(allocations) =>
+      LOG.info(s"Resource allocated for appMaster $appId")
+      val ResourceAllocation(resource, worker, workerId) = allocations(0)
+      val appMasterContext = AppMasterContext(appId, username, executorId, resource, jar, null, AppMasterRuntimeInfo(worker))
+
+      LOG.info(s"Try to launch a executor for app Master on ${worker} for app $appId")
+      val name = ActorUtil.actorNameForExecutor(appId, executorId)
+      val selfPath = ActorUtil.getFullPath(context.system, self.path)
+
+      val jvmSetting = Util.resolveJvmSetting(app.conf, systemConfig).appMater
+      val executorJVM = ExecutorJVMConfig(jvmSetting.classPath ,jvmSetting.vmargs,
+        classOf[ActorSystemBooter].getName, Array(name, selfPath), jar, username)
+
+      worker ! LaunchExecutor(appId, executorId, resource, executorJVM)
+      context.become(waitForActorSystemToStart(worker, appMasterContext, app.conf))
+  }
+
+  def waitForActorSystemToStart(worker : ActorRef, appContext : AppMasterContext, user : UserConfig) : Receive = {
+    case ExecutorLaunchRejected(reason, resource, ex) =>
+      LOG.error(s"Executor Launch failed reason：$reason", ex)
+      LOG.info(s"reallocate resource $resource to start appmaster")
+      master ! RequestResource(appId, ResourceRequest(resource))
+      context.become(waitForResourceAllocation)
```

Comment:
will this become an infinite loop? executor launch rejected -> resource allocatioin -> executor launch rejected ....

may be add a counter for retry better.